de.fu_berlin.inf.dpp.core/src/de/fu_berlin/inf/dpp/session/internal/SharedReferencePointMapper.java

```diff
@@ -0,0 +1,537 @@
+     * at all
+     */
+    public synchronized void addResources(IReferencePoint referencePoint,
+        Collection<? extends IResource> resources) {
+
+        if (referencePointToIDMapping.get(referencePoint) == null) {
+            LOG.warn("could not add resources to referencePoint "
+                + referencePoint + " because it is not shared");
+            return;
+        }
+
+        if (completelySharedReferencePoints.contains(referencePoint)) {
+            LOG.warn("cannot add resources to completely shared referencePoint: "
+                + referencePoint);
+            return;
+        }
+
+        Set<IResource> partiallySharedResources = partiallySharedResourceMapping
+            .get(referencePoint);
+
+        if (partiallySharedResources.isEmpty()) {
+            partiallySharedResources = new HashSet<IResource>(Math.max(1024,
+                (resources.size() * 3) / 2));
+
+            partiallySharedResourceMapping.put(referencePoint,
+                partiallySharedResources);
+        }
+
+        partiallySharedResources.addAll(resources);
+    }
+
+    /**
+     * Removes the given resources from a <b>partially</b> shared reference
+     * point.
+     * 
+     * @param referencePoint
+     *            a reference point that was added as a partially shared
+     *            reference point
+     * @param resources
+     *            the resources to remove
+     */
+    /*
+     * TODO needs proper sync. in the SarosSession class
+     * 
+     * @throws IllegalStateException if the referencePoint is completely or not
+     * shared at all
+     */
+    public synchronized void removeResources(IReferencePoint referencePoint,
+        Collection<? extends IResource> resources) {
+
+        if (referencePointToIDMapping.get(referencePoint) == null) {
+            LOG.warn("could not remove resources from referencePoint "
+                + referencePoint + " because it is not shared");
+            return;
+        }
+
+        if (completelySharedReferencePoints.contains(referencePoint)) {
+            LOG.warn("cannot remove resources from completely shared reference point: "
+                + referencePoint);
+            return;
+        }
+
+        Set<IResource> partiallySharedResources = partiallySharedResourceMapping
+            .get(referencePoint);
+
+        partiallySharedResources.removeAll(resources);
+    }
+
+    /**
+     * Atomically removes and adds resources. The resources to remove will be
+     * removed first before the resources to add will be added.
+     * 
+     * @param referencePoint
+     *            a reference point that was added as a partially shared
+     *            reference point
+     * @param resourcesToRemove
+     *            the resources to remove
+     * @param resourcesToAdd
+     *            the resources to add
+     */
+    /*
+     * TODO needs proper sync. in the SarosSession class
```

Comment:
![Code Smell](https://sonarsource.github.io/sonar-github/code_smell.png 'Code Smell') Code Smell: Complete the task associated to this TODO comment. ([squid:S1135](https://sonarcloud.io/organizations/saros/rules#rule_key=squid%3AS1135))

[See it in SonarCloud](https://sonarcloud.io/project/issues?id=saros_complete&pullRequest=211&issues=AWVYURY9iZQY_aXE9GJP&open=AWVYURY9iZQY_aXE9GJP)

[ISSUE_KEY]: AWVYURY9iZQY_aXE9GJP